flink-tests/src/test/java/org/apache/flink/test/checkpointing/RescalingITCase.java

```diff
@@ -290,6 +301,139 @@ public void testSavepointRescalingKeyedState(boolean scaleOut, boolean deriveMax
         }
     }
 
+    @Test
+    public void testCheckpointRescalingInKeyedState() throws Exception {
+        testCheckpointRescalingKeyedState(false, false);
+    }
+
+    @Test
+    public void testCheckpointRescalingOutKeyedState() throws Exception {
+        testCheckpointRescalingKeyedState(true, false);
+    }
+
+    @Test
+    public void testCheckpointRescalingInKeyedStateDerivedMaxParallelism() throws Exception {
+        testCheckpointRescalingKeyedState(false, true);
+    }
+
+    @Test
+    public void testCheckpointRescalingOutKeyedStateDerivedMaxParallelism() throws Exception {
+        testCheckpointRescalingKeyedState(true, true);
+    }
+
+    /**
+     * Tests that a job with purely keyed state can be restarted from a checkpoint with a different
+     * parallelism.
+     */
+    public void testCheckpointRescalingKeyedState(boolean scaleOut, boolean deriveMaxParallelism)
```

Comment:
I believe most of code could be shared with `testSavepointRescalingKeyedState`, let's simplify the logic.